src/pykeen/pipeline.py

```diff
@@ -998,12 +998,28 @@ def pipeline(  # noqa: C901
 
     # Train like Cristiano Ronaldo
     training_start_time = time.time()
-    losses = training_loop_instance.train(
-        stopper=stopper,
-        result_tracker=result_tracker,
-        clear_optimizer=clear_optimizer,
-        **training_kwargs,
-    )
+    while True:
+        try:
+            losses = training_loop_instance.train(
+                stopper=stopper,
+                result_tracker=result_tracker,
+                clear_optimizer=clear_optimizer,
+                **training_kwargs,
+            )
+        except (MemoryError, RuntimeError) as e:
+            # If the training failed due to OOM on the GPU, the training is reverted to CPU training.
+            if training_loop_instance.device.type == 'cuda':
+                logging.warning("Tried to train the current model on GPU, but the model is too big for the GPU")
+                logging.warning("Reverting to CPU now, which will increase the training time significantly.")
+                training_loop_instance.model._set_device('cpu')
+                training_loop_instance.model.regularizer.device = torch.device('cpu')
+                training_loop_instance.model.reset_parameters_()
```

Comment:
@lvermue we should be more explicit about resetting the model's parameters (=starting a new training).

Also, does this take care of resetting the optimizer's parameters/states?

Reply:
An alternative could be to move the current parameters / gradients / states to cpu, and not reset the parameters to continue the same training on cpu. 

A use-case might be, when e.g. a second user accidentally starts a process on "my GPU", leading to OOM.

Reply:
Isn't the pipeline by default resetting the parameters/starting a new training?

Reply:
Good catch with the optimizer. I'll add the re-initiation of the optimizer.

> An alternative could be to move the current parameters / gradients / states to cpu, and not reset the parameters to continue the same training on cpu.
A use-case might be, when e.g. a second user accidentally starts a process on "my GPU", leading to OOM.

To continue training we would need to change the architecture of the training loop, since the inner state of the training loop is not accessible to the pipeline currently and that's where the errors are handled. Furthermore, we can't handle optimizers with moments, since we have to re-init them (see https://github.com/pytorch/pytorch/pull/3463#issuecomment-341673802). This means that we couldn't guarantee that the results are not altered.

Reply:
> Isn't the pipeline by default resetting the parameters/starting a new training?

Yes, but what is the benefit of starting a new pipeline on cpu compared to "just" crashing, and having the user to restart on cpu?

Reply:
I really think we should have it crash and force the user to update the settings rather than doing lots of possibly opaque internal fixes

Reply:
The benefit is that one can run the pipeline unattended with the exact configurations one has set. Currently, especially in the case of HPO, those runs not fitting on the GPU aren't performed. In cases where one is interested in exactly those configurations, regardless whether it fits on GPU or not, the way to deal with it right now is very cumbersome, because it is unknown what combination of parameters works on the GPU at hand and which ones don't.

Reply:
I suppose the more common use case is to run as many different configurations as possible in a given window of time. Falling back to cpu in such case might block the search for the complete time with one huge model.

At least we should enable to get the old behaviour of having crashing runs for GPU OOM models (which will just be dropped from HPO).